### PR TITLE
Init

### PR DESCRIPTION
**Pull request by CMORPHY**

### CODE REVIEW

File: src/sonar-project.ts [hunk 1]
```diff
@@ -84,14 +84,17 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
     //   throw new Error('Unimplemented')
     // },
     async uploadSettings () {
+      /* We can't set these settings via the Web API */
+      const rejectedSettings = (setting) => !['$schema', 'sonar.organization'].includes(setting)
```
Comment (dschnare) [inline src/sonar-project.ts]: Add `projectKey` to this check.

File: schemas/project.json [hunk 2]
```diff
@@ -15,17 +15,161 @@
       "type": "string",
       "pattern": "^[a-z0-9_.:\\-]+$"
     },
-    "sonar.projectDescription": {
+    "sonar.projectName": {
       "description": "Name of the project that will be displayed on the web interface.",
       "type": "string"
     },
+    "sonar.projectVersion": {
```
Comment (dschnare) [inline schemas/project.json]: Did you add `sonar.test.inclusions`? That would be typed array of strings. Does `sonar.sources` and `sonar.tests` need to be typed array of string as well?

Comment (CMORPHY) [reply schemas/project.json]: That makes sense, I'll do that now. We'll always use [ "str" ] for a single string as well.
Unless the Schema can support both types: a string and an array of strings.

Comment (dschnare) [reply schemas/project.json]: Yeah, we'll have to test these props to see if "comma separated" actually means we need to pass them as `values=value&values=value2` over the wire or not. If so, then these have to be array of stings in the JSON.

**Comment by dschnare** [comment]
In fact it might be good to look at the the settings in the web UI. We may have missed a few.

File: schemas/project.json [hunk 3]
```diff
@@ -15,17 +15,161 @@
...
+      "description": "The project version.",
+      "type": "string"
+    },
+    "sonar.login": {
```
Comment (dschnare) [inline schemas/project.json]: I would remove `sonar.login` and `sonar.password`. These are not ever going to be stored on the SonarCloud project. We only support access token based auth for this CLI tool.

File: schemas/project.json [hunk 4]
```diff
@@ -15,17 +15,161 @@
...
+    },
+    "sonar.links.scm": {
+      "description": "Project source repository.",
+      "type": "string"
+    },
+    "sonar.sourceEncoding": {
+      "description": "Encoding of the source files. Ex: UTF-8, MacRoman, Shift_JIS",
+      "type": "string"
+    },
+    "sonar.externalIssuesReportPaths": {
```
Comment (dschnare) [inline schemas/project.json]: Another one that *may* have to be array of strings. We'll know once you test setting `sonar.sources` as `values=value1`

File: schemas/project.json [hunk 5]
```diff
@@ -15,17 +15,161 @@
...
+      "description": "Comma-delimited list of paths to Generic Issue reports.",
+      "type": "string"
+    },
+    "sonar.projectDate": {
+      "description": "Assign a date to the analysis. This parameter is only useful when you need to retroactively create the history of a not-analyzed-before project. The format is yyyy-MM-dd, for example: 2010-12-01.",
+      "type": "string"
```
Comment (dschnare) [inline schemas/project.json]: Add a JSON format key here: `format: "date"`.
See: https://json-schema.org/understanding-json-schema/reference/string.html

File: schemas/project.json (hunk 3 above)
Comment (CMORPHY) [reply schemas/project.json]: Agreed, I followed the Web UI links though, but I'll double-check again if any are missing.

**Comment by CMORPHY** [comment]
> In fact it might be good to look at the the settings in the web UI. We may have missed a few.

Will double-check

**Comment by CMORPHY** [comment]
OK all addressed now - please take a look.
I've used type: [ "array", "string" ] for certain types that can be single or multiple values.

File: src/sonar-project.ts [hunk 6]
```diff
@@ -84,14 +108,17 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
     //   throw new Error('Unimplemented')
     // },
     async uploadSettings () {
+      /* We can't set these settings via the Web API */
+      const rejectedSettings = (setting) => !['$schema', 'sonar.organization', 'sonar.projectKey'].includes(setting)
       const setEndpoint = `${BASE_URL}/api/settings/set`
       const resetEndpoint = `${BASE_URL}/api/settings/reset`
       const settingKeys = Object.keys(settings)
```
Comment (dschnare) [inline src/sonar-project.ts]: Change this to: `const settingKeys = [...settings.keys()]`

File: src/sonar-project.ts [hunk 7]
```diff
@@ -84,14 +108,17 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
...
-      const resettingSettingKeys = settingKeys.filter(key => settings[key] === null)
-      const settingSettingKeys = settingKeys.filter(key => settings[key] !== null)
+      const resettingSettingKeys = settingKeys.filter(key => settings[key] === null).filter(rejectedSettings)
```
Comment (dschnare) [inline src/sonar-project.ts]: Change this to: `const resettingSettingKeys = settingKeys.filter(key => settings.get(key) === null).filter(rejectedSettings)`

File: src/sonar-project.ts [hunk 8]
```diff
@@ -84,14 +108,17 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
...
+      const settingSettingKeys = settingKeys.filter(key => settings[key] !== null).filter(rejectedSettings)
```
Comment (dschnare) [inline src/sonar-project.ts]: Change this to: `const settingSettingKeys = settingKeys.filter(key => settings.get(key) !== null).filter(rejectedSettings)`

File: src/sonar-project.ts [hunk 9]
```diff
@@ -100,7 +127,7 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
             `keys=${resettingSettingKeys.join(',')}`
           ],
           accessToken
-        }),
+        }) : null,
         settingSettingKeys.map(key => {
           const value = settings[key]
```
Comment (dschnare) [inline src/sonar-project.ts]: Chance to `const value = settings.get(key)`

File: src/sonar-project.ts [hunk 10]
```diff
@@ -58,20 +70,32 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
...
+  })
+
   return {
     get settings () {
-      return settings as any
+      return settings
+
     },
     set settings (value) {
       if (Object(value) === value && !Array.isArray(value)) {
```
Comment (dschnare) [inline src/sonar-project.ts]: Change this to test that the value being set is an instance of Map:
```
if (value instanceof Map) {
  settings = new Map([...value])
} else {
  throw ...
}

```

File: src/sonar-project.ts [hunk 11]
```diff
@@ -58,20 +70,32 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
...
-    defaultSettings,
-    await parseJsonWithComments(
-      projectRepo.readFile('.sonar-project.json', { encoding: 'utf8' }).catch(error => {
-        if (error.code === 'ENOENT') return '{}'
-        throw error
-      })
-    )
+
+  Object.keys(defaultSettings).forEach((key) => {
+    settings[key] = {
```
Comment (dschnare) [inline src/sonar-project.ts]: Change this to
```
settings.set(key, {
  value: defaultSettings[key],
  isFromDefaults: true
})
```

File: src/sonar-project.ts [hunk 12]
```diff
@@ -58,20 +70,32 @@ export async function SonarProject (options: SonarProjectOptions): Promise<Sonar
...
+
+  const settingsFromRepo = await parseJsonWithComments(
+    projectRepo.readFile('.sonar-project.json', { encoding: 'utf8' }).catch(error => {
+      if (error.code === 'ENOENT') return '{}'
+      throw error
+    })
   )
 
+  Object.keys(settingsFromRepo).forEach((key) => {
+    settings[key] = {
```
Comment (dschnare) [inline src/sonar-project.ts]: Change this to:
```
settings.set(key, {
  value: settingsFromRepo[key],
  isFromDefaults: false
})
```

**Comment by dschnare** [comment]
You can have a look at the Map API here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

File: src/sonar-project.ts (hunk 11 above)
Comment (dschnare) [reply src/sonar-project.ts]: Don't forget handling the default settings properly. Have to call Map's `set()` method.

Comment (CMORPHY) [reply src/sonar-project.ts]: @dschnare yesterday as I was going to sleep, I remembered that we have used settings[key]=..at two places :) seriously, I was like.. when I wake up, gotta use .set() there too, it only makes sense :) :)

**Pull request merged**